### PR TITLE
feat(relay): finish nakama migration from starter to world engine

### PR DESCRIPTION
**Pull request by pyrofolium**
Does not close #462

requires additional PR to starter-game-template to complete

## What is the purpose of the change

Upgrade world-engine nakama so it has all the features from the nakama in starter and all the tests as well. 

## Testing and Verifying

There's a new test to test out the event system.

### CODE REVIEW

File: internal/e2e/tester/cardinal/sys/join.go [hunk 1]
```diff
@@ -26,6 +28,7 @@ func Join(ctx cardinal.WorldContext) error {
 		}
 		PlayerEntityID[jtx.Sig().PersonaTag] = entityID
 		logger.Info().Msgf("player %s successfully joined", jtx.Sig().PersonaTag)
+		ctx.EmitEvent(fmt.Sprintf("%d player: %d created, %d/%d", entityID, entityID, entityID, 3))
```
Comment (technicallyty) [inline internal/e2e/tester/cardinal/sys/join.go:31]: - not sure we need the "3" at the end
- do we need to list the entity id 3 times?

File: internal/test_nakama/util.go [hunk 2]
```diff
@@ -30,6 +32,65 @@ func newClient(t *testing.T) *nakamaClient {
 	return h
 }
 
+type NotificationItem struct {
+	ID         string    `json:"id"`
+	Subject    string    `json:"subject"`
+	Content    string    `json:"content"`
+	Code       int       `json:"code"`
+	CreateTime time.Time `json:"create_time"`
```
Comment (technicallyty) [inline internal/test_nakama/util.go:40]:
```suggestion
	CreateTime time.Time `json:"createTime"`
```

File: internal/test_nakama/util.go [hunk 3]
```diff
@@ -30,6 +32,65 @@ func newClient(t *testing.T) *nakamaClient {
...
+	Persistent bool      `json:"persistent"`
+}
+
+type Content struct {
+	Message string `json:"message"`
+}
+
+type NotificationCollection struct {
+	Notifications   []NotificationItem `json:"notifications"`
+	CacheableCursor string             `json:"cacheable_cursor"`
```
Comment (technicallyty) [inline internal/test_nakama/util.go:50]:
```suggestion
	CacheableCursor string             `json:"cacheableCursor"`
```

File: relay/nakama/main.go [hunk 4]
```diff
@@ -122,6 +126,32 @@ func initReceiptDispatcher(log runtime.Logger) error {
 	return nil
 }
 
+func initEventHub(ctx context.Context, log runtime.Logger, nk runtime.NakamaModule) error {
+	eventHub, err := createEventHub(log)
+	if err != nil {
+		return err
+	}
+	go func() {
+		_ = eventHub.dispatch(log)
```
Comment (technicallyty) [inline relay/nakama/main.go:135]:
```suggestion
		err := eventHub.dispatch(log)
```

File: internal/e2e/tester/cardinal/sys/join.go (hunk 1 above)
Comment (pyrofolium) [reply internal/e2e/tester/cardinal/sys/join.go:31]: hmm this is for a test. Previously I could actually get the amount transactions. Now with this new method there's no access to the length. I was thinking that in the future I could create a method that could access the length and replace it here but I think I'm just going to modify the test.

Comment (technicallyty) [reply internal/e2e/tester/cardinal/sys/join.go]: saw the fix, but still seems unnecessary to send the entity ID 3 times here, we can just simplify to 1?

Comment (pyrofolium) [reply internal/e2e/tester/cardinal/sys/join.go]: Well, this test actually caught one of the huge bugs I was trying to work on. Basically events were lossy, if I sent events too fast I would actually be missing events. If I didn't send 3 this bug would not have been caught.

Comment (technicallyty) [reply internal/e2e/tester/cardinal/sys/join.go]: missing event data or missing entire events? my point is that you have 
```go
ctx.EmitEvent(fmt.Sprintf("%d player: %d created, %d/%d", entityID, entityID, entityID, 3))
```
when you could just do 
```go
ctx.EmitEvent(fmt.Sprintf("player entity id %d created", entityID))
```

Comment (pyrofolium) [reply internal/e2e/tester/cardinal/sys/join.go]: I meant Missing entire events. 

Ok I thought you meant emitting the id, three times not the string itself. 

Yeah, idc, The test, however requires the first value to be the entityid. So I'll just make it work with that.

**Review by technicallyty — APPROVED** [review]

**Pull request merged**